HTMLScriptElement.text - trustedTypes for scripts (#41128)
mdn/content: files/en-us/web/api/htmlscriptelement/text/index.md

{{APIRef("HTML DOM")}}

> [!WARNING]
> This property represents the text content of a script element, which may be executable depending on the script type.

> > APIs like this are known as [injection sinks](/en-US/docs/Web/API/Trusted_Types_API#concepts_and_usage), and are potentially a vector for [cross-site-scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks.

Author: Note, slightly different warning as this isn't DOM injection. Also because a script might not be executable.

> See [Security considerations](#security_considerations) for more information.

The **`text`** property of the {{domxref("HTMLScriptElement")}} interface represents the inline text content of the script element.
It acts the same way as the {{domxref("HTMLScriptElement.textContent","textContent")}} property.

It reflects the `text` attribute of the {{HTMLElement("script")}} element.

## Value

Getting the property returns a string containing the element's text.

Setting the property accepts either a {{domxref("TrustedScript")}} object or a string.

## Description

The **`text`** property of the {{domxref("HTMLScriptElement")}} interface represents the text content inside the {{HTMLElement("script")}} element.

For an executable script {{domxref('HTMLScriptElement/type','type')}}, such as a module or classic script, this text is inline executable code.
For other types it might represent an import map, speculation rules, or some other kind of data block.

Note that if the {{domxref('HTMLScriptElement/src','src')}} property is set the content of the `text` property is ignored.

> ### text vs textContent vs innerText

Reviewer: ```suggestion
### `text` vs `textContent` vs `innerText`
```

it's ugly but I think good practice to do this, especially for the benefit of localizers.

The `text` and {{domxref("HTMLScriptElement.textContent", "textContent")}} properties of `HTMLScriptElement` are equivalent: both can be set with a string or a `TrustedScript` type and both return a string representing the content of the script element.
The main difference is that {{domxref("Node.textContent", "textContent")}} is also defined on {{domxref("Node")}} and can be used with other elements to set their content with a string.

> {{domxref("HTMLScriptElement.innerText", "innerText")}} will generally set and execute the text in the same as the other methods, but may return a slightly different value.

Reviewer: ```suggestion
{{domxref("HTMLScriptElement.innerText", "innerText")}} will generally set and execute the text in the same way as the other methods, but may return a slightly different value.
```

This does not change the execution of the text, but it does alter the text that is stored and returned.

### Security considerations

The `text` property is a possible vector for [Cross-site-scripting (XSS)](/en-US/docs/Web/Security/Attacks/XSS) attacks, where potentially unsafe strings provided by a user are executed.
For example, the following example assumes the `scriptElement` is an executable `<script>` element, and that `untrustedCode` was provided by a user:

```js
const untrustedCode = "alert('Potentially evil code!');";
scriptElement.text = untrustedCode; // shows the alert
```

You can mitigate these issues by always assigning {{domxref("TrustedScript")}} objects instead of strings, and [enforcing trusted type](/en-US/docs/Web/API/Trusted_Types_API#using_a_csp_to_enforce_trusted_types) using the [`require-trusted-types-for`](/en-US/docs/Web/HTTP/Reference/Headers/Content-Security-Policy/require-trusted-types-for) CSP directive.
This ensures that the input is passed through a transformation function, which has the chance to [sanitize](/en-US/docs/Web/Security/Attacks/XSS#sanitization) or reject the text before it is injected.

> The behavior of the transformation function will depend on the specific use case that requires a user provided script.

Author: Wasn't entirely sure what to say here. The fact is that you shouldn't generally be taking untrusted strings and injecting them directly into scripts. If you do then using trustedScripts is a good idea, but it is impossible to provide general guidance for something that you would only be doing in extreme circumstances. I've tried to lock it down a bit and give some flavour of the risk in the avoidance suggestions.

If possible you should lock the allowed scripts to exactly the code that you trust to run.

> If that is not possible, you might whitelist or blacklist the use of certain functions within the provided string.

Reviewer: ```suggestion
If that is not possible, you might allow or block the use of certain functions within the provided string.
```

? maybe? -> https://developers.google.com/style/word-list#blacklist

For the purpose of this example we'll allow just exactly the script that we need.

```js
const policy = trustedTypes.createPolicy("inline-script-policy", {
  createScript: (input) => {
    // Here specify what scripts are safe to allow
    if (input === "const num = 10;\nconsole.log(num)") {
      return input; // allow this exact script
    }
    throw new TypeError("Untrusted script blocked: " + input);
  },
});
```

Next we'll create the script element to which we will assign the value and get a handle to the element.

```html
<script id="el" type="text/javascript"></script>
```

```js
// Get the script element we're injecting the code into
const el = document.getElementById("el");
```

Then we use the `policy` object to create a `trustedScript` object from the potentially unsafe input string, and assign the result to the element:

```js
// The potentially malicious string
const untrustedScriptOne = "const num = 10;\nconsole.log(num)";

// Create a TrustedScript instance using the policy
const trustedScript = policy.createScript(untrustedScriptOne);

// Inject the TrustedScript (which contains a trusted string)
el.text = trustedScript;
```

> ### text and textContent properties are equivalent

Reviewer: ```suggestion
### Comparing `text` and `textContent`
```

idk why, but headings whose text tells me things irk me.
